🐛 fix: Development server shows white screen with error logger runtime errors

## Overview

The development server is displaying a white screen instead of the application, accompanied by error logger runtime errors in the console. This blocks local development and testing.

## Problem Statement

When running the development server (`npm run dev`), users encounter:
- **White screen** instead of the expected application UI
- **Console errors** related to error logging and hook systems
- **Application failure** to render any content

## Console Error Details

```
Global error handler initialized successfully
errorLogger.ts:444 [ERROR LOGGER] RUNTIME
hook.js:608 Error: Object
overrideMethod @ hook.js:608Understand this error
errorLogger.ts:446 Context: Object
```

## Reproduction Steps

1. Clone the repository
2. Install dependencies with `npm install`
3. Start development server with `npm run dev`
4. Navigate to `http://localhost:3001`
5. Observe white screen and console errors

## Technical Analysis

**Error Logger Analysis:**
- Error occurs at `errorLogger.ts:444` - the console logging function
- Line 444: `console.group(\`%c[ERROR LOGGER] ${error.category.toUpperCase()}\`, style)`
- Line 446: `console.log('Context:', context)`
- Suggests an error object is being logged that contains circular references or problematic data

**Hook System Analysis:**
- Error originates from `hook.js:608` in node_modules dependencies
- Likely related to React style systems or side effect management
- Multiple potential sources:
  - `react-style-singleton/dist/*/hook.js`
  - `use-sidecar/dist/*/hook.js`

**Root Cause Hypothesis:**
1. **Circular Reference Error**: Error context contains circular references causing JSON.stringify issues
2. **Hook Integration Error**: React hooks in error handling system causing initialization failures
3. **Global Error Handler Recursion**: Error handler catching its own errors, creating infinite loop
4. **Development vs Production Mismatch**: Environment-specific error handling causing conflicts

## Environment Details

**Development Configuration:**
- Node.js: (check with `node --version`)
- React: 19.1.0
- Vite: 6.3.5
- Error Logger: Custom implementation at `src/utils/errorLogger.ts:444`
- Global Error Handler: `src/utils/globalErrorHandler.ts`

**Error Handler Stack:**
```
main.tsx:10 → initializeGlobalErrorHandling()
App.tsx:564 → ErrorBoundary wrapper
App.tsx:46 → useErrorHandler('ExamContent')
```

## Proposed Investigation Steps

### Phase 1: Error Logger Debugging
- [ ] Add try-catch around `console.group` and `console.log` calls in error logger
- [ ] Implement circular reference detection for error context logging
- [ ] Add debug mode to error logger with simplified logging
- [ ] Check if error context contains problematic objects (DOM elements, React components)

### Phase 2: Hook System Analysis
- [ ] Identify which hook.js file is causing the error (use browser dev tools source maps)
- [ ] Check for conflicts between React style libraries
- [ ] Verify React Strict Mode compatibility with error handling
- [ ] Test with error logging temporarily disabled

### Phase 3: Global Error Handler Review
- [ ] Implement error recursion detection in global handler
- [ ] Add safeguards against error handler generating errors
- [ ] Create fallback error display that bypasses complex logging
- [ ] Test error boundary isolation

## Acceptance Criteria

- [ ] Development server loads without white screen
- [ ] Application renders correctly in development mode
- [ ] Error logging works without causing application crashes
- [ ] Console shows only expected initialization messages
- [ ] Error handling system functions without recursion

## Workaround Options

### Immediate Fixes
1. **Disable Error Logging**: Temporarily comment out error logger initialization
2. **Simplify Error Context**: Remove complex objects from error context
3. **Basic Error Display**: Replace complex ErrorBoundary with simple error display
4. **Production Mode**: Test with production build to isolate development-specific issues

### Configuration Changes
```javascript
// Temporary fix in main.tsx
// initializeGlobalErrorHandling({
//   showUserNotifications: false,
//   attemptAutoRecovery: false,
// })
```

## Development Impact

**Blocks:**
- Local development and testing
- Component development workflow
- Error handling verification
- Accessibility testing with axe-core

**Affects:**
- Developer productivity
- Debugging capabilities
- Error monitoring system
- Component error boundaries

## Files Involved

**Primary Files:**
- `src/utils/errorLogger.ts:444` - Error logging console output
- `src/utils/globalErrorHandler.ts` - Global error handling initialization
- `src/main.tsx:10` - Error handler initialization
- `src/components/ui/ErrorBoundary.tsx` - React error boundaries

**Dependencies:**
- `node_modules/react-style-singleton/dist/*/hook.js`
- `node_modules/use-sidecar/dist/*/hook.js`
- React 19.1.0 error handling APIs

## References

- Error Logger implementation: `src/utils/errorLogger.ts:441-448`
- Global error handler: `src/utils/globalErrorHandler.ts:124-148`
- Error boundary usage: `src/App.tsx:564-601`
- Main entry point: `src/main.tsx:6-23`

--- a/src/components/ui/ErrorBoundary.tsx
+++ b/src/components/ui/ErrorBoundary.tsx
@@ -57,11 +57,19 @@ export class ErrorBoundary extends Component<Props, State> {
 
   componentDidCatch(error: Error, errorInfo: ErrorInfo) {
     const componentName = this.props.componentName || 'UnknownComponent'
+    let errorId: string | null = null
 
-    // Log error to global error handling system
-    const errorId = logRuntimeError(error, componentName, 'ErrorBoundary', {
-      session: this.getSessionInfo(),
-    })
+    try {
+      // Log error to global error handling system with safe context
+      errorId = logRuntimeError(error, componentName, 'ErrorBoundary', {
+        session: this.getSessionInfo(),
+      })
+    } catch (loggingError) {
+      // If error logging fails, use fallback logging
+      console.error('[ErrorBoundary] Error logging failed:', loggingError)
+      console.error('[ErrorBoundary] Original error:', error.message)
+      errorId = `fallback-${Date.now()}`
+    }
 
     this.setState({
       error,
@@ -69,21 +77,35 @@ export class ErrorBoundary extends Component<Props, State> {
       errorId,
     })
 
-    // Call the error callback if provided
-    this.props.onError?.(error, errorInfo)
+    try {
+      // Call the error callback if provided
+      this.props.onError?.(error, errorInfo)
+    } catch (callbackError) {
+      console.error('[ErrorBoundary] Error callback failed:', callbackError)
+    }
 
     // Attempt automatic recovery if enabled and not too many retries
     if (this.props.enableAutoRecovery && this.state.retryCount < 3) {
-      this.scheduleAutoRecovery()
+      try {
+        this.scheduleAutoRecovery()
+      } catch (recoveryError) {
+        console.error('[ErrorBoundary] Auto recovery setup failed:', recoveryError)
+      }
     }
 
-    // Console logging for development
+    // Safe console logging for development
     if (process.env.NODE_ENV === 'development') {
-      console.group(`🚨 ErrorBoundary: ${componentName}`)
-      console.error('Error:', error)
-      console.error('Error Info:', errorInfo)
-      console.error('Component Stack:', errorInfo.componentStack)
-      console.groupEnd()
+      try {
+        console.group(`🚨 ErrorBoundary: ${componentName}`)
+        console.error('Error:', error.message || error)
+        console.error('Error Info:', {
+          componentStack: errorInfo.componentStack?.substring(0, 500) + '...' || 'No stack available'
+        })
+        console.groupEnd()
+      } catch (consoleError) {
+        console.error('[ErrorBoundary] Console logging failed:', consoleError)
+        console.error('[ErrorBoundary] Original error was:', error.message)
+      }
     }
   }
 
@@ -134,15 +156,22 @@ export class ErrorBoundary extends Component<Props, State> {
       if (sessionData) {
         const parsed = JSON.parse(sessionData)
         return {
-          sessionId: parsed.sessionId || 'unknown',
-          duration: Date.now() - (parsed.testState?.startTime || Date.now()),
-          isActive: !parsed.testState?.isCompleted,
+          sessionId: String(parsed.sessionId || 'unknown').substring(0, 50),
+          duration: Number(Date.now() - (parsed.testState?.startTime || Date.now())) || 0,
+          isActive: Boolean(!parsed.testState?.isCompleted),
         }
       }
-    } catch {
-      // Ignore parsing errors
+    } catch (sessionError) {
+      // Ignore parsing errors but log in development
+      if (process.env.NODE_ENV === 'development') {
+        console.warn('[ErrorBoundary] Failed to get session info:', sessionError)
+      }
     }
-    return undefined
+    return {
+      sessionId: 'session-unavailable',
+      duration: 0,
+      isActive: false,
+    }
   }
 
   render() {

--- a/src/main.tsx
+++ b/src/main.tsx
@@ -6,37 +6,63 @@ import './styles/index.css'
 // Initialize global error handling
 import { initializeGlobalErrorHandling } from './utils/globalErrorHandler'
 
-// Initialize global error handling before anything else
-initializeGlobalErrorHandling({
-  showUserNotifications: true,
-  attemptAutoRecovery: true,
-  maxErrorsBeforeWarning: 3,
-  errorCountingWindow: 30000, // 30 seconds
-  notificationPreferences: {
-    showNotifications: true,
-    notificationTimeout: 8000, // 8 seconds
-    showTechnicalDetails: process.env.NODE_ENV === 'development',
-    position: 'top',
-    soundEnabled: false,
-    vibrationEnabled: false,
-  },
-})
+// Initialize global error handling before anything else with safe error handling
+try {
+  initializeGlobalErrorHandling({
+    showUserNotifications: true,
+    attemptAutoRecovery: true,
+    maxErrorsBeforeWarning: 3,
+    errorCountingWindow: 30000, // 30 seconds
+    notificationPreferences: {
+      showNotifications: true,
+      notificationTimeout: 8000, // 8 seconds
+      showTechnicalDetails: process.env.NODE_ENV === 'development',
+      position: 'top',
+      soundEnabled: false,
+      vibrationEnabled: false,
+    },
+  })
+  console.log('Global error handling initialized successfully')
+} catch (initError) {
+  console.error('Failed to initialize global error handling:', initError)
+  // Continue with application startup even if error handling fails
+}
 
 // Initialize memory management system
 import { initializeMemoryManagement } from './utils/memoryCleanup'
-initializeMemoryManagement()
+try {
+  initializeMemoryManagement()
+  console.log('Memory management initialized')
+} catch (error) {
+  console.warn('Failed to initialize memory management:', error)
+}
 
 // Initialize question preloading for better performance
 import { initializeQuestionPreloading } from './utils/questionLoader'
-initializeQuestionPreloading()
+try {
+  initializeQuestionPreloading()
+  console.log('Question preloading initialized')
+} catch (error) {
+  console.warn('Failed to initialize question preloading:', error)
+}
 
 // Initialize performance monitoring
 import { initializePerformanceMonitoring } from './utils/performanceMonitoring'
-initializePerformanceMonitoring()
+try {
+  initializePerformanceMonitoring()
+  console.log('Performance monitoring initialized')
+} catch (error) {
+  console.warn('Failed to initialize performance monitoring:', error)
+}
 
 // Initialize caching system
 import { initializeCaching } from './utils/cacheStrategies'
-initializeCaching()
+try {
+  initializeCaching()
+  console.log('Caching system initialized')
+} catch (error) {
+  console.warn('Failed to initialize caching system:', error)
+}
 
 // Initialize browser compatibility detection early
 import { initializeBrowserCompatibility } from './utils/browserCompatibility'

--- a/src/utils/errorLogger.ts
+++ b/src/utils/errorLogger.ts
@@ -361,43 +361,161 @@ export class ErrorLogger {
   private createFullContext(
     partialContext: Partial<ErrorContext>
   ): ErrorContext {
-    const fullContext: ErrorContext = {
-      environment: {
-        userAgent: navigator.userAgent,
-        language: navigator.language,
-        screenSize: {
-          width: window.screen.width,
-          height: window.screen.height,
+    try {
+      const fullContext: ErrorContext = {
+        environment: {
+          userAgent: navigator.userAgent?.substring(0, 200) || 'Unknown',
+          language: navigator.language || 'Unknown',
+          screenSize: {
+            width: window.screen?.width || 0,
+            height: window.screen?.height || 0,
+          },
+          timezone: this.getSafeTimezone(),
+          cookieEnabled: Boolean(navigator.cookieEnabled),
+          localStorage: this.checkStorageAvailable('localStorage'),
+          sessionStorage: this.checkStorageAvailable('sessionStorage'),
         },
-        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
-        cookieEnabled: navigator.cookieEnabled,
-        localStorage: this.checkStorageAvailable('localStorage'),
-        sessionStorage: this.checkStorageAvailable('sessionStorage'),
-      },
-      ...partialContext,
-    }
+        ...this.sanitizePartialContext(partialContext),
+      }
 
-    // Add performance information if available
-    if (window.performance) {
-      fullContext.performance = {
-        memoryUsage: (performance as any).memory?.usedJSHeapSize || 0,
-        timing: performance.timing,
-        navigation: performance.navigation,
+      // Add performance information if available (safe extraction)
+      if (window.performance) {
+        fullContext.performance = {
+          memoryUsage: this.getSafeMemoryUsage(),
+          timing: this.getSafePerformanceTiming(),
+          navigation: this.getSafeNavigationInfo(),
+        }
+      }
+
+      // Add network information if available (safe extraction)
+      if ('connection' in navigator) {
+        const connection = (navigator as any).connection
+        if (connection) {
+          fullContext.network = {
+            quality: this.determineNetworkQuality(connection),
+            effectiveType: connection.effectiveType || 'unknown',
+            downlink: Number(connection.downlink) || 0,
+            rtt: Number(connection.rtt) || 0,
+          }
+        }
+      }
+
+      return fullContext
+    } catch (contextError) {
+      // Return minimal context if full context creation fails
+      console.warn('Failed to create full error context:', contextError)
+      return {
+        environment: {
+          userAgent: 'Context creation failed',
+          language: 'unknown',
+          screenSize: { width: 0, height: 0 },
+          timezone: 'unknown',
+          cookieEnabled: false,
+          localStorage: false,
+          sessionStorage: false,
+        },
+        session: {
+          sessionId: 'context-creation-failed',
+          duration: 0,
+          isActive: false,
+        },
       }
     }
+  }
 
-    // Add network information if available
-    if ('connection' in navigator) {
-      const connection = (navigator as any).connection
-      fullContext.network = {
-        quality: this.determineNetworkQuality(connection),
-        effectiveType: connection.effectiveType,
-        downlink: connection.downlink,
-        rtt: connection.rtt,
-      }
+  /**
+   * Sanitize partial context to remove complex objects
+   */
+  private sanitizePartialContext(partialContext: Partial<ErrorContext>): Partial<ErrorContext> {
+    if (!partialContext || typeof partialContext !== 'object') {
+      return {}
     }
 
-    return fullContext
+    const sanitized: Partial<ErrorContext> = {}
+
+    try {
+      // Only include safe, primitive properties
+      for (const [key, value] of Object.entries(partialContext)) {
+        if (value === null || value === undefined) {
+          sanitized[key as keyof ErrorContext] = value
+        } else if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
+          sanitized[key as keyof ErrorContext] = value
+        } else if (key === 'session' && typeof value === 'object') {
+          // Safe session data extraction
+          const sessionData = value as any
+          sanitized.session = {
+            sessionId: String(sessionData.sessionId || 'unknown').substring(0, 50),
+            duration: Number(sessionData.duration) || 0,
+            isActive: Boolean(sessionData.isActive),
+          }
+        } else {
+          // Skip complex objects
+          sanitized[key as keyof ErrorContext] = `[${typeof value}]` as any
+        }
+      }
+    } catch (err) {
+      console.warn('Failed to sanitize partial context:', err)
+    }
+
+    return sanitized
+  }
+
+  /**
+   * Get safe timezone information
+   */
+  private getSafeTimezone(): string {
+    try {
+      return Intl.DateTimeFormat().resolvedOptions().timeZone || 'Unknown'
+    } catch {
+      return 'Unknown'
+    }
+  }
+
+  /**
+   * Get safe memory usage information
+   */
+  private getSafeMemoryUsage(): number {
+    try {
+      return (performance as any).memory?.usedJSHeapSize || 0
+    } catch {
+      return 0
+    }
+  }
+
+  /**
+   * Get safe performance timing information
+   */
+  private getSafePerformanceTiming(): any {
+    try {
+      const timing = performance.timing
+      if (!timing) return undefined
+
+      // Extract only essential timing info to avoid circular references
+      return {
+        navigationStart: timing.navigationStart || 0,
+        loadEventEnd: timing.loadEventEnd || 0,
+        domContentLoadedEventEnd: timing.domContentLoadedEventEnd || 0,
+      }
+    } catch {
+      return undefined
+    }
+  }
+
+  /**
+   * Get safe navigation information
+   */
+  private getSafeNavigationInfo(): any {
+    try {
+      const navigation = performance.navigation
+      if (!navigation) return undefined
+
+      return {
+        type: navigation.type || 0,
+        redirectCount: navigation.redirectCount || 0,
+      }
+    } catch {
+      return undefined
+    }
   }
 
   /**
@@ -439,12 +557,169 @@ export class ErrorLogger {
    * Log error to console in development
    */
   private logToConsole(error: ApplicationError, context: ErrorContext): void {
-    const style = this.getConsoleStyle(error.severity)
+    try {
+      const style = this.getConsoleStyle(error.severity)
 
-    console.group(`%c[ERROR LOGGER] ${error.category.toUpperCase()}`, style)
-    console.error('Error:', error)
-    console.log('Context:', context)
-    console.groupEnd()
+      console.group(`%c[ERROR LOGGER] ${error.category.toUpperCase()}`, style)
+      
+      // Safely log error with circular reference handling
+      try {
+        console.error('Error:', this.sanitizeForLogging(error))
+      } catch (errorLogErr) {
+        console.error('Error (sanitized):', {
+          id: error.id,
+          message: error.message,
+          category: error.category,
+          severity: error.severity,
+          timestamp: error.timestamp
+        })
+      }
+      
+      // Safely log context with circular reference handling
+      try {
+        console.log('Context:', this.sanitizeForLogging(context))
+      } catch (contextLogErr) {
+        console.log('Context (sanitized):', this.createSafeContext(context))
+      }
+      
+      console.groupEnd()
+    } catch (consoleError) {
+      // Fallback logging if all else fails
+      console.error('[ERROR LOGGER] Failed to log error safely:', error.message || 'Unknown error')
+    }
+  }
+
+  /**
+   * Sanitize objects for safe console logging (handles circular references)
+   */
+  private sanitizeForLogging(obj: any, maxDepth = 3): any {
+    const seen = new WeakSet()
+    
+    const sanitize = (value: any, depth: number): any => {
+      // Depth limit to prevent infinite recursion
+      if (depth > maxDepth) {
+        return '[Max Depth Reached]'
+      }
+
+      // Handle null and undefined
+      if (value === null || value === undefined) {
+        return value
+      }
+
+      // Handle primitives
+      if (typeof value !== 'object') {
+        return value
+      }
+
+      // Handle circular references
+      if (seen.has(value)) {
+        return '[Circular Reference]'
+      }
+      seen.add(value)
+
+      // Handle arrays
+      if (Array.isArray(value)) {
+        return value.slice(0, 10).map(item => sanitize(item, depth + 1)) // Limit array size
+      }
+
+      // Handle DOM elements
+      if (value instanceof Element) {
+        return `[DOM Element: ${value.tagName}${value.id ? '#' + value.id : ''}]`
+      }
+
+      // Handle functions
+      if (typeof value === 'function') {
+        return `[Function: ${value.name || 'anonymous'}]`
+      }
+
+      // Handle Error objects
+      if (value instanceof Error) {
+        return {
+          name: value.name,
+          message: value.message,
+          stack: value.stack?.split('\n').slice(0, 5).join('\n') // Limit stack trace
+        }
+      }
+
+      // Handle React components/elements
+      if (value.$$typeof || (value._owner !== undefined) || (value.type !== undefined)) {
+        return '[React Element/Component]'
+      }
+
+      // Handle plain objects
+      if (value.constructor === Object || value.constructor === undefined) {
+        const sanitized: any = {}
+        const keys = Object.keys(value).slice(0, 20) // Limit object keys
+        
+        for (const key of keys) {
+          try {
+            sanitized[key] = sanitize(value[key], depth + 1)
+          } catch (err) {
+            sanitized[key] = '[Error accessing property]'
+          }
+        }
+        
+        if (Object.keys(value).length > 20) {
+          sanitized['...'] = `[${Object.keys(value).length - 20} more properties]`
+        }
+        
+        return sanitized
+      }
+
+      // For other object types, try to extract basic info
+      try {
+        return `[Object: ${value.constructor?.name || 'Unknown'}]`
+      } catch {
+        return '[Complex Object]'
+      }
+    }
+
+    try {
+      return sanitize(obj, 0)
+    } catch (err) {
+      return `[Sanitization Failed: ${err instanceof Error ? err.message : 'Unknown error'}]`
+    }
+  }
+
+  /**
+   * Create a safe context object with only essential properties
+   */
+  private createSafeContext(context: ErrorContext): any {
+    try {
+      return {
+        environment: {
+          userAgent: context.environment?.userAgent?.substring(0, 100) || 'Unknown',
+          language: context.environment?.language || 'Unknown',
+          screenSize: context.environment?.screenSize || { width: 0, height: 0 },
+          timezone: context.environment?.timezone || 'Unknown',
+          cookieEnabled: Boolean(context.environment?.cookieEnabled),
+          localStorage: Boolean(context.environment?.localStorage),
+          sessionStorage: Boolean(context.environment?.sessionStorage),
+        },
+        performance: context.performance ? {
+          memoryUsage: context.performance.memoryUsage || 0,
+          timingAvailable: Boolean(context.performance.timing),
+          navigationAvailable: Boolean(context.performance.navigation),
+        } : undefined,
+        network: context.network ? {
+          quality: context.network.quality || 'unknown',
+          effectiveType: context.network.effectiveType || 'unknown',
+          downlink: context.network.downlink || 0,
+          rtt: context.network.rtt || 0,
+        } : undefined,
+        session: context.session || undefined,
+        // Exclude component and boundary as they're not part of ErrorContext type
+        // These are handled in the ApplicationError interface itself
+      }
+    } catch (err) {
+      return {
+        session: {
+          sessionId: 'safe-context-failed',
+          duration: 0,
+          isActive: false,
+        },
+      }
+    }
   }
 
   /**

--- a/src/utils/globalErrorHandler.ts
+++ b/src/utils/globalErrorHandler.ts
@@ -113,6 +113,10 @@ export class GlobalErrorHandler {
   private originalErrorHandler?: OnErrorEventHandler
   private originalRejectionHandler?: (event: PromiseRejectionEvent) => void
   private notificationContainer?: HTMLElement
+  private isHandlingError = false
+  private recursionDepth = 0
+  private readonly MAX_RECURSION_DEPTH = 3
+  private readonly RECURSION_RESET_DELAY = 1000
 
   constructor(config: Partial<GlobalErrorHandlerConfig> = {}) {
     this.config = { ...DEFAULT_CONFIG, ...config }
@@ -185,9 +189,25 @@ export class GlobalErrorHandler {
     error: ApplicationError,
     context: any = {}
   ): Promise<ErrorHandlerResult> {
+    // Check for recursion before any processing
+    if (this.isHandlingError || this.recursionDepth >= this.MAX_RECURSION_DEPTH) {
+      console.warn('[GlobalErrorHandler] Recursion detected, using fallback handling')
+      return this.getFallbackErrorResult(error)
+    }
+
+    // Set recursion protection
+    this.isHandlingError = true
+    this.recursionDepth++
+
+    // Reset recursion depth after delay
+    setTimeout(() => {
+      this.recursionDepth = Math.max(0, this.recursionDepth - 1)
+    }, this.RECURSION_RESET_DELAY)
+
     try {
-      // Log the error
-      const errorKey = errorLogger.logError(error, context)
+      // Log the error with safe context
+      const safeContext = this.createSafeErrorContext(context)
+      const errorKey = errorLogger.logError(error, safeContext)
 
       // Update error tracking
       this.updateErrorTracking(error)
@@ -229,13 +249,55 @@ export class GlobalErrorHandler {
       return recoveryResult
     } catch (handlingError) {
       console.error('Error while handling error:', handlingError)
-      return {
-        handled: false,
-        actionsTaken: [],
-        requiresUserAction: true,
-        userMessage: 'Radās neparedzēta kļūda. Lūdzu, pārlādējiet lapu.',
-        shouldContinue: false,
+      return this.getFallbackErrorResult(error)
+    } finally {
+      // Always reset handling flag
+      this.isHandlingError = false
+    }
+  }
+
+  /**
+   * Create safe error context by removing problematic properties
+   */
+  private createSafeErrorContext(context: any): any {
+    if (!context || typeof context !== 'object') {
+      return context
+    }
+
+    try {
+      const safeContext: any = {}
+      
+      // Only include safe, primitive properties
+      for (const [key, value] of Object.entries(context)) {
+        if (value === null || value === undefined) {
+          safeContext[key] = value
+        } else if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
+          safeContext[key] = value
+        } else if (key === 'source' || key === 'userAgent' || key === 'timestamp') {
+          safeContext[key] = String(value).substring(0, 200) // Truncate long strings
+        } else {
+          // Skip complex objects that might cause circular references
+          safeContext[key] = `[${typeof value}]`
+        }
       }
+      
+      return safeContext
+    } catch (err) {
+      return { error: 'Failed to create safe context' }
+    }
+  }
+
+  /**
+   * Get fallback error result for critical failures
+   */
+  private getFallbackErrorResult(error: ApplicationError): ErrorHandlerResult {
+    return {
+      handled: false,
+      actionsTaken: [],
+      requiresUserAction: true,
+      userMessage: 'Radās neparedzēta kļūda. Lūdzu, pārlādējiet lapu.',
+      technicalMessage: error.message || 'Unknown error',
+      shouldContinue: false,
     }
   }
 
@@ -247,42 +309,55 @@ export class GlobalErrorHandler {
     this.originalErrorHandler = window.onerror
 
     window.onerror = (message, source, lineno, colno, error) => {
-      // Call original handler first
-      if (this.originalErrorHandler) {
-        this.originalErrorHandler(message, source, lineno, colno, error)
-      }
+      try {
+        // Call original handler first
+        if (this.originalErrorHandler) {
+          this.originalErrorHandler(message, source, lineno, colno, error)
+        }
 
-      // Create error object
-      const runtimeError: ApplicationError = {
-        id: `window-error-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
-        message: typeof message === 'string' ? message : 'Nezināma kļūda',
-        details: error?.stack || `${source}:${lineno}:${colno}`,
-        severity: 'error',
-        category: 'runtime',
-        timestamp: Date.now(),
-        recoverable: true,
-        suggestedActions: ['refresh', 'retry'],
-        context: {
-          source,
-          line: lineno,
-          column: colno,
-        },
-        originalError:
-          error || new Error(message?.toString() || 'Unknown error'),
-        stack: error?.stack,
-      }
+        // Skip handling if we're already in a recursion loop
+        if (this.isHandlingError || this.recursionDepth >= this.MAX_RECURSION_DEPTH) {
+          console.warn('[GlobalErrorHandler] Skipping window.onerror due to recursion')
+          return true
+        }
 
-      // Handle the error asynchronously
-      setTimeout(() => {
-        this.handleError(runtimeError, {
-          source: 'window.onerror',
-          userAgent: navigator.userAgent,
+        // Create error object
+        const runtimeError: ApplicationError = {
+          id: `window-error-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
+          message: typeof message === 'string' ? message : 'Nezināma kļūda',
+          details: error?.stack || `${source}:${lineno}:${colno}`,
+          severity: 'error',
+          category: 'runtime',
           timestamp: Date.now(),
-        })
-      }, 0)
+          recoverable: true,
+          suggestedActions: ['refresh', 'retry'],
+          context: {
+            source,
+            line: lineno,
+            column: colno,
+          },
+          originalError:
+            error || new Error(message?.toString() || 'Unknown error'),
+          stack: error?.stack,
+        }
 
-      // Return true to prevent default error handling
-      return true
+        // Handle the error asynchronously
+        setTimeout(() => {
+          this.handleError(runtimeError, {
+            source: 'window.onerror',
+            userAgent: navigator.userAgent?.substring(0, 100),
+            timestamp: Date.now(),
+          }).catch(handlingError => {
+            console.error('[GlobalErrorHandler] Failed to handle window error:', handlingError)
+          })
+        }, 0)
+
+        // Return true to prevent default error handling
+        return true
+      } catch (setupError) {
+        console.error('[GlobalErrorHandler] Error in window.onerror handler:', setupError)
+        return true
+      }
     }
   }
 
@@ -291,48 +366,60 @@ export class GlobalErrorHandler {
    */
   private setupPromiseRejectionHandler(): void {
     const handler = (event: PromiseRejectionEvent) => {
-      // Prevent default unhandled rejection logging
-      event.preventDefault()
+      try {
+        // Prevent default unhandled rejection logging
+        event.preventDefault()
 
-      const reason = event.reason
-      let error: ApplicationError
+        // Skip handling if we're already in a recursion loop
+        if (this.isHandlingError || this.recursionDepth >= this.MAX_RECURSION_DEPTH) {
+          console.warn('[GlobalErrorHandler] Skipping promise rejection due to recursion')
+          return
+        }
 
-      if (reason instanceof Error) {
-        error = {
-          id: `promise-rejection-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
-          message: reason.message || 'Neapstrādāta Promise noraidīšana',
-          details: reason.stack,
-          severity: 'error',
-          category: 'runtime',
-          timestamp: Date.now(),
-          recoverable: true,
-          suggestedActions: ['retry', 'refresh'],
-          originalError: reason,
-          stack: reason.stack,
+        const reason = event.reason
+        let error: ApplicationError
+
+        if (reason instanceof Error) {
+          error = {
+            id: `promise-rejection-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
+            message: reason.message || 'Neapstrādāta Promise noraidīšana',
+            details: reason.stack,
+            severity: 'error',
+            category: 'runtime',
+            timestamp: Date.now(),
+            recoverable: true,
+            suggestedActions: ['retry', 'refresh'],
+            originalError: reason,
+            stack: reason.stack,
+          }
+        } else {
+          // Handle non-Error rejections
+          error = {
+            id: `promise-rejection-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
+            message: 'Neapstrādāta Promise noraidīšana',
+            details: String(reason).substring(0, 200), // Limit string length
+            severity: 'warning',
+            category: 'runtime',
+            timestamp: Date.now(),
+            recoverable: true,
+            suggestedActions: ['retry'],
+            context: { rejectionReason: typeof reason },
+          }
         }
-      } else {
-        // Handle non-Error rejections
-        error = {
-          id: `promise-rejection-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
-          message: 'Neapstrādāta Promise noraidīšana',
-          details: String(reason),
-          severity: 'warning',
-          category: 'runtime',
-          timestamp: Date.now(),
-          recoverable: true,
-          suggestedActions: ['retry'],
-          context: { rejectionReason: reason },
-        }
+
+        // Handle the error asynchronously
+        setTimeout(() => {
+          this.handleError(error, {
+            source: 'unhandledrejection',
+            promiseReason: typeof reason,
+            timestamp: Date.now(),
+          }).catch(handlingError => {
+            console.error('[GlobalErrorHandler] Failed to handle promise rejection:', handlingError)
+          })
+        }, 0)
+      } catch (setupError) {
+        console.error('[GlobalErrorHandler] Error in promise rejection handler:', setupError)
       }
-
-      // Handle the error asynchronously
-      setTimeout(() => {
-        this.handleError(error, {
-          source: 'unhandledrejection',
-          promiseReason: reason,
-          timestamp: Date.now(),
-        })
-      }, 0)
     }
 
     this.originalRejectionHandler = handler

